feat: dashboard 기능 구현

uha9218 opened the pull request:
https://www.notion.so/N-1-23f0f19ac33780f99814e96301332c8e 적용

uha9218 committed:
feat: Vue 기본 설정 및 Dashboard 목업 코드 작성

uha9218 committed:
feat : ScheduleCalender 기능 개발

uha9218 committed:
feat: H2 테스트 환경 설정 및 JPA 매핑 문제 해결

- H2 메모리 DB를 테스트 환경으로 설정
- JPA 엔티티 매핑 문제 해결 (Schedule-Section 양방향 관계)
- PostgreSQL 드라이버 로딩 테스트 추가
- ScheduleCalendarServiceIntegrationTest H2 환경에서 정상 동작 확인
- 메인 애플리케이션은 PostgreSQL, 테스트는 H2로 분리

uha9218 committed:
feat: 달력 기능 프론트엔드 연동

- 백엔드: ScheduleCalendarController 추가 (월별 일정 조회 API)
- 프론트엔드: scheduleApi.js 서비스 파일 생성
- Dashboard.vue 컴포넌트 API 연동

uha9218 committed:
feat: 학습 진행 현황 기능 구현 및 현재 진행 중인 스터디 자동 조회 기능 추가

- 학습 진행 현황 API 구현 (StudyProgressController, StudyProgressService)
- 현재 진행 중인 스터디 자동 조회 기능 추가 (Study.isActive 필드)
- Schedule.hasDeepStudy, Section.needsReview 필드 추가
- 프론트엔드 학습 진행 현황 UI 구현 (Dashboard.vue)
- 현재 진행 중인 스터디가 없을 때 적절한 에러 처리 추가
- 관련 DTO, Repository, Service, Test 코드 모두 구현
- CORS 설정 추가

f-lab-lyan commented on src/main/java/com/example/soup/schedule/entity/Schedule.java:
```diff
@@ -50,6 +53,7 @@ public static Schedule create(
 		String description,
 		LocalDateTime scheduleDate,
 		String meetingLocation,
+		Boolean hasDeepStudy,
```
```suggestion
		boolean hasDeepStudy,
```

uha9218 committed:
feat: 성능 모니터링 및 테스트 환경 구축

- 성능 모니터링을 위한 AOP 로깅 기능 추가 (LogExecutionTime, ExecutionTimeLogger)
- JMeter를 이용한 부하 테스트 환경 구축
- P6Spy 설정으로 SQL 쿼리 모니터링 추가
- 스케줄 캘린더 관련 통합 테스트 및 부하 테스트 추가
- 테스트 실행을 위한 배치 파일 추가
- 성능 모니터링 가이드 문서 추가

uha9218 committed:
feat: StudyProgressService 테스 클래스 분리 및 N+1 쿼리 문제 검증 테스트 추가

f-lab-lyan commented on src/main/java/com/example/soup/SoupApplication.java:
```diff
@@ -10,4 +13,17 @@ public static void main(String[] args) {
 		SpringApplication.run(SoupApplication.class, args);
 	}
 
+	@Bean
```
이 빈은 여기말고 다른 config클래스에 넣는 것이 좋지 않을까요?

uha9218 committed:
test: Study Progress N+1성능 측정용 테스트 추가가

f-lab-lyan commented on src/test/java/com/example/soup/PostgresDriverTest.java:
```diff
@@ -0,0 +1,11 @@
+
+    @Test
+    void canLoadPgDriver() throws Exception {
+        Class.forName("org.postgresql.Driver"); // 여기서 예외 나면 classpath 문제
```
이건 클래스패스에 postgresql이 포함되어 있는지 확인하는 테스트 인 것인가요?!

f-lab-lyan commented on bash.exe.stackdump:
이 파일은 왜 포함된 것인가요! 😄

f-lab-lyan commented on tatus:
오 .. 이 파일은 정체가 무엇인가요??

f-lab-lyan commented on src/test/java/com/example/soup/schedule/ScheduleCalendarServiceUnitTest.java:
```diff
@@ -0,0 +1,236 @@
+
+    @BeforeEach
+    void setUp() {
+        scheduleRepository = mock(ScheduleRepository.class);
```
@mock @injectMock 😀

f-lab-lyan commented on src/main/java/com/example/soup/study/service/StudyProgressService.java:
```diff
@@ -0,0 +1,107 @@
+    private final StudyRepository studyRepository;
+    private final ScheduleRepository scheduleRepository;
+    private final SectionRepository sectionRepository;
+    private final UserRepository userRepository;
```
이것보다는 UserService를 포함하는 것이 더 좋지 않았을까하는 생각이 드네요. DB를 감춘다는 측면에서요.

f-lab-lyan commented on src/main/java/com/example/soup/study/controller/StudyProgressController.java:
```diff
@@ -0,0 +1,50 @@
+                    .build();
+            StudyProgressResponseDTO response = studyProgressService.getStudyProgress(request);
+            return ResponseEntity.ok(response);
+        } catch (IllegalArgumentException e) {
```
ExceptionHandler등을 도입해 보는게 좋을 것 같습니다.

f-lab-lyan commented on src/main/java/com/example/soup/schedule/service/ScheduleCalendarService.java:
```diff
@@ -0,0 +1,66 @@
+        long endTime = System.currentTimeMillis();
+        
+        log.info("데이터베이스 조회 완료: {}개 일정, 소요시간: {}ms", schedules.size(), endTime - startTime);
+        
```
CustomException?

f-lab-lyan reviewed (approved):
수고하셨습니다. 👍 몇몇 파일들을 지울 필요가 있어보이는데, 그 부분 빼고는 다 괜찮은 것 같습니다. 😄

uha9218 committed:
feat: JPA Fetch Join을 활용한 N+1 문제 해결 및 성능 최적화

- StudyRepository에 Fetch Join 쿼리 추가 (findByIdWithDetails, findByIsActiveTrueWithDetails)
- StudyProgressService 리팩토링: 단일 쿼리로 모든 연관 데이터 조회
- 엔티티 관계 개선: Study.schedules, Schedule.deepStudies, Section.reviews를 Set으로 변경
- MultipleBagFetchException 해결: 동시 Fetch Join 가능하도록 컬렉션 타입 최적화
- 단위 테스트 및 성능 테스트 수정: Set 컬렉션 변경에 따른 테스트 로직 업데이트
- 성능 검증: 대용량 데이터(12,500개 리뷰, 240개 딥스터디)에서 282ms 처리 성능 달성
- N+1 문제 완전 해결: 단일 Fetch Join 쿼리로 모든 연관 엔티티 조회

uha9218 committed:
refactor: StudyProgressService를 2-Step Inquiry 패턴으로 리팩토링

- 성능 최적화: 단일 스케줄 조회에 최적화된 2-Step Inquiry 패턴 적용
- 새로운 리포지토리 메서드 추가:
  - SectionRepository.findByScheduleIdWithReviews(): 스케줄별 섹션과 리뷰를 효율적으로 조회
  - DeepStudyRepository.findByScheduleIdWithUser(): 스케줄별 심화학습과 사용자를 효율적으로 조회
- StudyProgressRequestDTO에 scheduleId 필드 추가
- StudyProgressService.getStudyProgress() 메서드 시그니처 변경 (studyId, scheduleId)
- 서비스 로직 개선: Map 기반 O(1) 조회로 성능 향상
- 모든 단위 테스트 및 성능 테스트 업데이트
- 메모리 효율성 향상: 필요한 데이터만 정확히 조회
- 확장성 개선: UI에서 단일 스케줄 진행 현황 조회에 최적화

uha9218 committed:
refactor: CORS 설정 분리 및 코드 정리

- CORS 설정을 SoupApplication에서 별도 CorsConfig 클래스로 분리 (단일 책임 원칙)
- Boolean -> boolean 타입 수정 (primitive 타입 사용)
- 불필요한 파일들 삭제 (README, 테스트 파일, 배치 파일 등)

f-lab-lyan commented on src/main/java/com/example/soup/study/repository/StudyRepository.java:
```diff
@@ -1,16 +1,39 @@
+     * 단일 쿼리로 조회하여 N+1 문제를 해결합니다. (Fetch Join 적용)
+     * 'DISTINCT'를 사용하여 카티전 곱으로 인해 발생하는 Study 루트 엔티티의 중복을 제거합니다.
+     */
+    @Query("SELECT DISTINCT s FROM Study s " +
```
👍
